src/lib/recurringPurposes.js

```diff
@@ -1,18 +1,248 @@
+  let closestStart = {},
+    closestEnd = {}
+  let minStartDistance = Number.MAX_SAFE_INTEGER,
+    minEndDistance = Number.MAX_SAFE_INTEGER
+  for (const contact of contacts) {
+    for (const address of contact.address) {
+      const addressCoordinates = {
+        lon: address.geo.geo[0],
+        lat: address.geo.geo[1]
+      }
+      const startPlaceDistance = geodesicDistance(
+        startCoordinates,
+        addressCoordinates
+      )
+
+      const endPlaceDistance = geodesicDistance(
+        endCoordinates,
+        addressCoordinates
+      )
+
+      if (startPlaceDistance < minStartDistance) {
+        closestStart = {
+          distance: startPlaceDistance,
+          contact,
+          address,
+          newCoordinates: startCoordinates
+        }
+        minStartDistance = startPlaceDistance
+      }
+      if (endPlaceDistance < minEndDistance) {
+        closestEnd = {
+          distance: endPlaceDistance,
+          contact,
+          address,
+          newCoordinates: endCoordinates
+        }
+        minEndDistance = endPlaceDistance
+      }
+    }
+  }
+  return { closestStart, closestEnd }
+}
+
+const setAddressContactRelationShip = ({
+  timeserie,
+  contact,
+  addressId,
+  relType
+}) => {
+  const newRel = {
+    [relType]: {
+      data: {
+        _id: contact._id,
+        _type: CONTACTS_DOCTYPE,
+        metadata: {
+          addressId
+        }
+      }
+    }
+  }
+  const timeserieWithRel = {
+    ...timeserie,
+    relationships: { ...(timeserie.relationships || {}), ...newRel }
+  }
+
+  return timeserieWithRel
+}
+
+const updateAddressCoordinates = async ({
+  client,
+  newCoordinates,
+  contact,
+  addressId
+}) => {
+  // Query contact to get latest rev
+  const newContact = await queryContactByDocId(client, contact._id)
+  let addressIdx = -1
+  for (let i = 0; newContact.address.length; i++) {
+    if (newContact.address[i].id === addressId) {
+      addressIdx = i
+      break
+    }
+  }
+  if (addressIdx < 0) {
+    log('warn', `No address ${addressId} found on contact ${contact._id}`)
+    return null
+  }
+
+  const currCoordinates = newContact.address[addressIdx].geo?.geo
+  const currCount = newContact.address[addressIdx].geo?.count || 0
+  const currSum = newContact.address[addressIdx].geo?.sum || [0, 0]
+  const newSumLon = currSum[0] + newCoordinates.lon
+  const newSumLat = currSum[1] + newCoordinates.lat
+
+  if (!currCoordinates) {
+    newContact.address[addressIdx].geo.geo = [
+      newCoordinates.lon,
+      newCoordinates.lat
+    ]
+  } else {
+    // XXX - This is an incremental average of longitudes and latitudes, that does not
+    // take into consideration of curvature of the Earth. Hence, this is an approximation.
+    // We consider this is acceptable as the coordinates are supposed to be very close, i.e. few meters.
+    // For more precise computations, one should typically use a geodesic center of all the points.
```

Comment:
s/one/we ?

Reply:
Why that? "one" seems  valid to me here:  https://en.wikipedia.org/wiki/One_(pronoun)